Framework: Implemented data-layer freshness verification.

jorgefilipecosta opened the pull request:
Data-layer freshness verification allows requesters to avoid requests that were recently done.
Automatic tests are not yet created, when we are confident that this is the path to go and this API's are ok, I will implement automatic tests.

Sample usage: https://github.com/Automattic/wp-calypso/pull/16729.

I changed our middleware to allow actions to be ignored /discarded avoiding them to arrive in the reducers this is required to avoid the reducer setting a isRequesting state when in fact it is not requesting and it would never leave that state.

dmsnell commented:
Thanks for this PR! I think it might make more sense to look into wrapping `initiator` with a function that can filter the freshness constraints. We don't need to make it part of the actual middleware in that sense and we can create the wrapper in a more-generic and probably more testable way.

Each call to `dispatchRequest()` will be unique as the code today stands. That means we can probably get away with simpler state tracking…

```js
/**
 * Abort a request if our data is fresh-enough
 *
 * @param {Number} id tracks the type of request being issued
 * @param {Function} initiator issues actual network requests
 * @param {Map} history tracks state of requests, injected for testing
 * @returns {Function} wrapped request initiator
 */
export const withFreshness = ( id, initiator, history = requestHistory ) => ( store, action, next ) => {
	const freshness = get( action, 'meta.dataLayer.fresherThan' );

	// if no freshness requested
	// then just issue it
	if ( ! freshness ) {
		return initiator( store, action, next );
	}

	const lastUpdate = history.get( id ) || -Infinity;
	const now = Date.now();

	const staleness = now - lastUpdate;

	// our data is fresher than we need it to be
	// so just skip this fetch
	if ( staleness <= freshness ) {
		return;
	}

	// otherwise, fetch it and update state
	history.set( id, now );
	return initiator( store, action, next );
}
```

now obviously this is incomplete and meant only to be an example. we probably ought to rather trap `onSuccess` than the `initiator` for updating the history. as in, the initiator should abort requests, but `onSuccess` should update the `lastUpdate` value (however and in whichever form it's stored)

```js
export const dispatchRequest = … => {
	// remember this specific interactor
	const id = _.uniqueId();
	const augmentedInitiator = withFreshness( id, initiator ); 

	return {
		…
		return augmentedInitiator( store, action, next );
	}
}
```

jorgefilipecosta commented:
Hi @dmsnell, thank you a lot for your fast review! I tried to incorporate some of your feedback and I changed some code. 

Regarding your suggestion for using a simple state tracking, e.g. using const id = _.uniqueId(); to generate history map keys, I think the map keys needs always to use information of the action being requested and not just the type of request e.g. actions { "COMMENTS_REQUEST" siteId: 128867501 postId: 97 … and  { "COMMENTS_REQUEST" siteId: 128867501 postId: 232… are the same request type and would have the same id so if dispatched consecutively the second is ignored but it should not because the request is different. I don’t think we can avoid adding action information to the key on our history map.

The second point is what to do when ignoring the action. We can not simply ignore it and let it arrive in the reducers because some reducers would set a isRequesting state and we would stay in that state for ever because no success or error would arrive. Right now I added a mechanism that allows our action to the discarded and not arrive in the reducers. The other option would and add an onIgnore option on dispatchRequest that dispatchRequest users could use to configure the dispatching of actions that clean isRequesting action. But that may force data-layer users to create new actions/reducers.

dmsnell commented:
> the map keys needs always to use information of the action being requested

yep. that's a proper detail that needs to be worked out. you can find a similar detail in the de-duper in the HTTP pipeline. we should rely heavily on HTTP semantics here to determine what can be safely ignored vs. what should be reasonably unsafe to ignore.

> some reducers would set a isRequesting state

this is unfortunate and counter to the patterns establed by the data layer, but it is what it is. would be better if we could remove those uses because `_REQUEST` is nothing more than a declaration "I need this resource" and _absolutely NOT_ an indication that a fetch is currently in progress.

pragmatically we should think about this. I want to be careful not to complicate the middleware any more than is necessary and make sure we explore other possible ways of preventing the passage of the action when we trap and ignore it.

dmsnell commented on client/state/data-layer/wpcom-api-middleware.js:
```diff
@@ -98,13 +98,14 @@ export const middleware = handlers => store => next => {
 		const safeNext = a => nextActions.add( a );
-
-		handlerChain.forEach( handler => handler( store, action, safeNext ) );
+		let shouldIgnore = false;
```
how _dare_ someone introduce a `let` into my middleware!

🤣 

I'm just teasing with you, but hopefully someone who reviews this will get a laugh at that.

I'm only half-joking though… haha

jorgefilipecosta replied:
Sorry :$ at the time it looked like the way with less code changes to get a discard mechanism, and I was thinking of better ways to do it. I will avoid hurting your middleware again :)

dmsnell commented on client/state/data-layer/wpcom-api-middleware.js:
```diff
@@ -98,13 +98,14 @@ export const middleware = handlers => store => next => {
-		handlerChain.forEach( handler => handler( store, action, safeNext ) );
+		let shouldIgnore = false;
+		const ignoreAction = () => shouldIgnore = true;
+		handlerChain.forEach( handler => handler( store, action, safeNext, ignoreAction ) );
```
adding another argument to the handler API is something I want to actively avoid. it may be plausible to use the return value as some kind of indicator, but even with that idea I'd like to ponder it for a bit before making a decision. we're in the process of _removing_ `next`/`safeNext` so that the handlers only get `store` and `action`

dmsnell commented on client/state/data-layer/wpcom-http/freshness-verifier.js:
```diff
@@ -0,0 +1,69 @@
+ */
+const buildActionKey = ( action ) => JSON.stringify(
+	action.meta ? { ...action, meta: undefined } : action
+);
```
this is probably too specific here. we're probably starting out only with `GET` requests because we have to be extra careful not to much around with `POST`s. further, a `GET` is fully specified by its path and query string. we should be able to build a key from a few properties. you can check in the pipeline with the dedupe request.

jorgefilipecosta replied:
Hi @dmsnell, Yes it is true that freshness is supposed to be used with gets, and gets can be defined by the path and query parameters, but on dispatchRequest we don't know yet how the action is mapped to path and query parameters. On the dedupe request we already have access to this mapping so the key build there can take advantage of that.

dmsnell replied:
or we send in a fake dispatcher and trap those HTTP calls, since it should dispatch an `HTTP_REQUEST` action. of course, then we couldn't prevent other non-HTTP initiating actions from taking place.

good point you make. 

> force users to pass an efficient key builder using what they know about the action

would love to make it so we don't have to think about this part. it's not something I think people should need to know about. it should "just work"

how about for now we leave in your key-builder which serializes the whole action without `meta` and continue to think about this problem

coderkevin commented:
> would be better if we could remove those uses because _REQUEST is nothing more than a declaration "I need this resource" and absolutely NOT an indication that a fetch is currently in progress.

@dmsnell, what would you then suggest for rendering some kind of a loading state? A rendered component which depends on data needs 3 display states:

1. data fulfilled and populated.
2. loading/waiting for data.
3. error state.

Prevailing wisdom suggests that we get that delineation of status from the redux state, correct? So, how should it be stored then for a given piece of data?

dmsnell commented:
> Prevailing wisdom suggests that we get that delineation of status from the redux state, correct? So, how should it be stored then for a given piece of data?

right @coderkevin. the ideal way is via some kind of nullity check on state. I think we _could_ generalize this pattern in such a way that we could all benefit from it without needing to do much coding (and store the timestamp of last refresh with the data somehow)

```js
const isLoading = state.resourceType[ resourceId ] === null; // if it's not in memory then it's loading
const failedToLoad = ?
```

so far signaling failures via notices has seemed to work well for a wide variety of cases. when we _have_ to show error status I'm less certain. probably it's best to encode that in the resource/data model itself. I like to compare it to what I would do in Elm…

```elm
type DooDad
  = NotAvailable String {- some string explanation, or a code, or whatever is necessary -}
  | DooDad { id : Number, description : String, … }
```

that is to say, we have two fundamental types that can occupy the state: a failure with an indication _or_ the actual resource with its properties. the _absence_ of this resource means that we're still waiting for it to load. the _success states_ indicate if we positively have a problem or if we positively have a success. _Most_ failures don't warrant state I think because we can just try again a couple times (already happening with `HTTP_REQUEST` actions) and notify the customer if we failed.

things like Woo which need more signaling that things positively failed probably ought to have some failure-type for the resource in state.

```js
{
	id: resourceId,
	lastUpdated: null,
	errorCode: …
}, {
	id: resourceId2,
	lastUpdated: Date.now(),
	description: 'blah blah blarg blurb',
	…
}
```

at least, this is where I think I'm leaning as far as design goes. I would really like to figure out if we could prevent the need for individual reducer authors from having to incorporate these meta fields.

in this sense, I think it's _intrinsic_ to Calypso that it's dealing with resources from some external storage. I don't think that _loading_ states are nominally part of that definition (that is they are _incidental_ because most of the ways we inject data into Calypso is from Calypso requesting that it be injected and then waiting).

long and fragile answer; did that address your prompt?

dmsnell commented:
@jorgefilipecosta what's the point of the cleaner? do we actually need it? why not keep that information around for the whole browser session? seems to be adding more code than I'd ideally like to see in this PR.

dmsnell commented:
also, we probably should have started with a README which explains the motivations for this feature, some of the background (such as the `<Interval />` component and the polling questions you were dealing with), how you think this should be used, some use cases, and some examples.

that would be 💯 😄

coderkevin commented:
> I think it's intrinsic to Calypso that it's dealing with resources from some external storage. I don't think that loading states are nominally part of that definition (that is they are incidental because most of the ways we inject data into Calypso is from Calypso requesting that it be injected and then waiting).

I agree that it's the Calypso Framework's responsibility to provide an easy-to-use way to retrieve resources with the goal of providing a consistent method to get that data. But given that it's the individual components of the application that must convey to the user some kind of appropriate loading state, or error state, I think it's important that the framework provide a consistent API or at least best practice around that part.

> I think we could generalize this pattern in such a way that we could all benefit from it without needing to do much coding (and store the timestamp of last refresh with the data somehow)

I'd like to chat with you about how this would look when you're available. But I think I'm done hijacking this PR thread for this purpose. 😆

samouri commented:
Different components could have different freshness constraints for the same resources.
Does this model support that?  I would think we would need to make new standard meta fields in the action object passed to the data-layer

samouri commented:
> @jorgefilipecosta what's the point of the cleaner? do we actually need it? why not keep that information around for the whole browser session? seems to be adding more code than I'd ideally like to see in this PR.

💯.  I agree with this, it is an over optimization for now.

samouri commented on client/state/data-layer/wpcom-http/utils.js:
```diff
@@ -66,28 +70,42 @@ export const getProgress = action => get( action, 'meta.dataLayer.progress', nul
+export const dispatchRequest = ( initiator, onSuccess, onError, onProgress = noop, options = {} ) => {
+	const { freshness } = options;
+	let initiatorToUse, onSuccessToUse;
+	if ( freshness ) {
```
i know it was suggested by @dmsnell, but this higher-order-initiator approach doesn't strike me as particularly clean

samouri commented:
the data-freshness problem has two sides to it:
1. making sure we don't send out more requests than we need to because data is fresh enough
2. automatically requesting data in the background when it has grown stale

This PR only addresses the former and I think keeping the latter in mind will help inform how we want to structure this code.  Not that I have a very clear idea of how.

Personally I was more a fan of adding the options object at the end with a `freshness` parameter whose shape could include things like: `maxAge`, and `shouldPoll`.  The higher-order-initiator approach seemed a bit funky to me.  We could even still write it similarly to how it is now by doing something like:
```js
export const dispatchRequest = ( initiator, onSuccess, onError, onProgress = noop, options = { freshness} ) => {
  initiator = initiatorWithFreshness( initiator, freshness); // identity function if no freshness specified
  onSuccess = onSuccessWithFreshness( onSuccess, freshness );
}
```


Separately, on the polling side, every N ms we could iterate over the objects in `history` and re-request things that are stale + should be polled.

dmsnell commented:
_Fresh_ idea: without abandoning this PR yet, could we create a new PR where @jorgefilipecosta you create a README for this new functionality and we can discuss the API there before writing any code?

jorgefilipecosta commented:
Sure @dmsnell, I will do that.

jorgefilipecosta commented:
Hi @dmsnell I created a new pull request as you suggested https://github.com/Automattic/wp-calypso/pull/16815 with a readme file https://github.com/Automattic/wp-calypso/tree/add/data-layer-freshness-verification-2/client/state/data-layer/freshness-verifier.


I will also, use this  comment to try to answer some of the points raised.

> @jorgefilipecosta what's the point of the cleaner? do we actually need it? why not keep that information around for the whole browser session? seems to be adding more code than I'd ideally like to see in this PR.

As our key is a serialization of the action and for each different action different request we save the key plus last time, I think our history may get big and waste lots of memory, so I tried to optimize that. But to be honest its just strings and a date so even with 1000 requests that would be not that much memory so yes I agree that was an over optimization and can be removed.


> Different components could have different freshness constraints for the same resources.
> Does this model support that?
@samouri,  For that may idea is to allow to pass the freshness in the action, I was not sure if that was needed so I did not implemented it at the start. The value set on the default handler could be default but the action may also set it. If some request should allow different freshness values that should be reflected in the action creator and that components could dispatch the freshness they need.

> Separately, on the polling side, every N ms we could iterate over the objects in history and re-request things that are stale + should be polled.
That approach has problem although a request may be in history the component that dispatched that requested may already be unmounted and the data not needed anymore.
My idea for polling would be to if it is really needed use the <Interval /> component to periodically dispatch request actions. So if the component is unmounted the polling stops and the data-layer does not need to worry about that.

dmsnell commented:
> My idea for polling would be to if it is really needed use the <Interval /> component to periodically dispatch request actions. So if the component is unmounted the polling stops and the data-layer does not need to worry about that.

I think could marry the `<Interval />` with query components. Imagine for example a `<QueryWidget />` that takes optional `fresherThan` and `keepUpToDate` properties…

```js
import { cond } from 'lodash'
import { EVERY_SECOND, …, EVERY_MINUTE } from 'lib/interval';

const updatePeriod = cond( [
	[ a => a <= 1000, () => EVERY_SECOND ],
	[ a => a <= 5000, () => EVERY_FIVE_SECONDS ],
	[ a => a <= 10000, () => EVERY_TEN_SECONDS ],
	…
	[ () => true, a => a / 1000 * 1000 ], // round to nearest second
] )

class QueryWidget extends Component {
	componentDidMount() {
		this.refresh();
	}

	componentDidUpdate( { id } ) {
		if ( id !== this.props.id ) {
			this.refresh();
		}
	}

	refresh = () => requestWidget( this.props.id );

	render() {
		if ( ! this.props.keepUpToDate ) {
			return null;
		}

		return <Interval onTick={ this.refresh } period={ updatePeriod( this.props.fresherThan ) } />;
	}
}

const mapDispatchToProps = ( {
	requestWidget: ( id, fresherThan ) => Object.assign(
		{
			type: WIDGET_REQUEST,
			id,
		},
		fresherThan && freshnessMeta( fresherThan )
	),
} )
```

just a thought

dmsnell commented:
> Separately, on the polling side, every N ms we could iterate over the objects in history and re-request things that are stale + should be polled.

well it's not hard to do a proper scheduled queue so I hope we wouldn't just blindly poll the events…

one common pattern is schedule the events and when the next one runs perform our cleanup

jorgefilipecosta commented:
> I think could marry the <Interval /> with query components. Imagine for example a <QueryWidget /> that takes optional fresherThan and keepUpToDate properties…

Exactly that's the idea I also have in mind instead of the query components just return null on their render they can if needed render the interval component, more or less equal to sample I tried here https://github.com/Automattic/wp-calypso/pull/16491/files for the query-post-counts case,  (there I used an identical new component because I was not aware of Interval existence.

apeatling commented:
@jorgefilipecosta What do you need to help get this PR merged?

jorgefilipecosta commented:
Given that there are discussions happening around this subject. And take 2 and other proposols like a metadata provider I will close this one for now. If because of some reason, we want soemething from here, I can easly revive it :)
